backends/tc/backend.cpp

```diff
@@ -0,0 +1,623 @@
+/*
+Copyright (C) 2023 Intel Corporation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions
+and limitations under the License.
+*/
+
+#include "backend.h"
+
+#include "backends/ebpf/ebpfOptions.h"
+#include "backends/ebpf/target.h"
+
+namespace TC {
+
+const cstring Extern::drop_packet = "drop_packet";
+const cstring Extern::send_to_port = "send_to_port";
+
+cstring PnaMainParserInputMetaFields[MAX_PNA_PARSER_META] = {"recirculated", "input_port"};
+
+cstring PnaMainInputMetaFields[MAX_PNA_INPUT_META] = {"recirculated", "timestamp", "parser_error",
+                                                      "class_of_service", "input_port"};
+
+cstring PnaMainOutputMetaFields[MAX_PNA_OUTPUT_META] = {"class_of_service"};
+
+const cstring pnaParserMeta = "pna_main_parser_input_metadata_t";
+const cstring pnaInputMeta = "pna_main_input_metadata_t";
+const cstring pnaOutputMeta = "pna_main_output_metadata_t";
+
+bool Backend::process() {
+    CHECK_NULL(toplevel);
+    if (toplevel->getMain() == nullptr) {
+        ::error("main is missing in the package");
+        return false;  //  no main
+    }
+    auto refMapEBPF = refMap;
+    auto typeMapEBPF = typeMap;
+    parseTCAnno = new ParseTCAnnotations();
+    tcIR = new ConvertToBackendIR(toplevel, pipeline, refMap, typeMap, options);
+    genIJ = new IntrospectionGenerator(pipeline, refMap, typeMap);
+    addPasses({parseTCAnno, new P4::ResolveReferences(refMap),
+               new P4::TypeInference(refMap, typeMap), tcIR, genIJ});
+    toplevel->getProgram()->apply(*this);
+    if (::errorCount() > 0) return false;
+    if (!ebpfCodeGen(refMapEBPF, typeMapEBPF)) return false;
+    return true;
+}
+
+bool Backend::ebpfCodeGen(P4::ReferenceMap *refMapEBPF, P4::TypeMap *typeMapEBPF) {
+    if (options.cFile.isNullOrEmpty()) return true;
+    target = new EBPF::KernelSamplesTarget(options.emitTraceMessages);
+    ebpfOption.emitTraceMessages = options.emitTraceMessages;
+    ebpfOption.maxTernaryMasks = options.maxTernaryMasks;
```

Comment:
Since P4TC doesn't use BPF maps to implement table lookups the following options should be removed as they are useless: maxTernaryMasks, enableTableCache. 

Also, `generateToXDP` should be removed as P4TC doesn't support XDP offload yet.

Reply:
removed the options